hpctestlib/microbenchmarks/mpi/osu.py

```diff
@@ -0,0 +1,168 @@
+            f'cd {self.build_prefix}'
+        ]
+        self.build_system.config_opts = [f'--enable-{self.build_type}']
+        self.build_system.make_opts = ['-C', 'mpi']
+        self.build_system.max_concurrency = 8
+
+    @sanity_function
+    def validate_build(self):
+        return True
+
+
+class osu_benchmark_test_base(rfm.RunOnlyRegressionTest):
+    '''Base class of OSU benchmarks runtime tests'''
+
+    #: Number of warmup iterations
+    #:
+    #: This value is passed to the excutable through the -x option.
+    #:
+    #: :type: :class:`int`
+    #: :default: ``1000``
+    num_warmup_iters = variable(int, value=1000)
+
+    #: Number of iterations
+    #:
+    #: This value is passed to the excutable through the -i option.
+    #:
+    #: :type: :class:`int`
+    #: :default: ``20000``
+    num_iters = variable(int, value=20000)
+
+    #: Maximum message size
+    #:
+    #: Both the performance and the sanity checks will be done
+    #: for this message size
+    #:
+    #: :type: :class:`int`
+    #: :default: ``8``
+    message_size = variable(int, value=8)
+
+    #: Device buffers
+    #:
+    #: Use accelerator device buffers, i.e cuda, openacc or rocm
+    #:
+    #: :type: :class:`str`
+    #: :default: ``None``
+    device_buffers = variable(str, type(None), value='cpu')
+
+    executable = ''
+    microbenchmarks = {
+        'collective': ['osu_alltoall', 'osu_allreduce'],
+        'pt2pt': ['osu_bw', 'osu_latency']
+    }
+    osu_binaries = fixture(build_osu_benchmarks, scope='environment')
+
+    @run_after('setup')
+    def set_executable(self):
+        if self.executable in self.microbenchmarks['collective']:
+            benchmark_type = 'collective'
+        elif self.executable in self.microbenchmarks['pt2pt']:
+            benchmark_type = 'pt2pt'
+
+        self.executable = os.path.join(
+            self.osu_binaries.stagedir, self.osu_binaries.build_prefix,
+            'mpi', benchmark_type, self.executable
+        )
+        max_message_size = max(self.message_size, self.message_size)
+        self.executable_opts = ['-m', f'{max_message_size}',
+                                '-x', f'{self.num_warmup_iters}',
+                                '-i', f'{self.num_iters}']
+        if self.device_buffers  and f'{self.device_buffers}' != 'cpu':
+            self.executable_opts += ['-d', f'{self.device_buffers}']
+
+            if benchmark_type == 'pt2pt':
+                self.executable_opts += ['D', 'D']
+
+    @sanity_function
+    def validate_test(self):
+        return sn.assert_found(rf'^{self.message_size}', self.stdout)
+
+
+class osu_bandwidth(osu_benchmark_test_base):
+    @performance_function('MB/s', perf_key='bw')
+    def bandwidth(self):
+        """Bandwidth for the message size `message_size`."""
+
+        return sn.extractsingle(rf'^{self.message_size}\s+(?P<bw>\S+)',
+                                self.stdout, 'bw', float)
+
+
+class osu_latency(osu_benchmark_test_base):
```

Comment:
The *only* difference of the `osu_latency` vs. `osu_bandwidth` is the `message_size`. The performance function should be moved to the base class and *not* use named capturing groups. There is no need for that and these should become as simple as the following:

```python
class osu_latency(osu_benchmark_test_base):
    message_size = 8

class osu_bandwidth(osu_benchmark_test_base):
    message_size = 1024*1024
```

I am also inclined to change the `osu_benchmark_test_base` to `osu_benchmark`.

Reply:
Correct point and I guess you look into different columns. I overlooked that.

Reply:
Correct point; I think you can still reuse the function. I'll have a look at it.